### PR TITLE
style: add clang format

### PR DESCRIPTION
**Pull request by gocarlos**
depends on https://github.com/cose-wg/COSE-C/pull/79

1. commit here: add clang format and format script
2. format source accordingly (automatically, no manual changes as impossible to review) -> this can be checked by getting first commit and running the format script -> should output the same as with the second commit applied

### CODE REVIEW

File: src/cose_int.h [hunk 1]
```diff
@@ -159,91 +167,227 @@ extern cose_error _MapFromCBOR(cn_cbor_errback err);
...
+    if (!(condition))                     \
+      {                                   \
+        DO_ASSERT;                        \
+        if (perr != NULL)                 \
+          {                               \
+            perr->err = error;            \
+          }                               \
+        goto errorReturn;                 \
+      }                                   \
+  }
```
Comment (gocarlos) [inline src/cose_int.h]: style is very subjective, though I think here the majority of people agree that this is more readable

File: src/cose_int.h [hunk 2]
```diff
@@ -159,91 +167,227 @@ extern cose_error _MapFromCBOR(cn_cbor_errback err);
...
-extern COSE_RecipientInfo * _COSE_Recipient_Init_From_Object(cn_cbor *, CBOR_CONTEXT_COMMA cose_errback * errp);
-extern void _COSE_Recipient_Free(COSE_RecipientInfo *);
-extern bool _COSE_Recipient_decrypt(COSE_RecipientInfo * pRecip, COSE_RecipientInfo * pRecipUse, int algIn, size_t cbitKey, byte * pbKey, cose_errback * errp);
-extern bool _COSE_Recipient_encrypt(COSE_RecipientInfo * pRecipient, const byte * pbContent, size_t cbContent, cose_errback * perr);
-extern byte * _COSE_RecipientInfo_generateKey(COSE_RecipientInfo * pRecipient, int algIn, size_t cbitKeySize, cose_errback * perr);
-
+extern bool _COSE_Init(COSE_INIT_FLAGS flags,
+                       COSE* pcose,
+                       int msgType,
+                       CBOR_CONTEXT_COMMA cose_errback* errp);
```
Comment (gocarlos) [inline src/cose_int.h]: style is very subjective, though I think here the majority of people agree that this is more readable

**Comment by gocarlos** [comment]
Failing as coveralls is annoyingly going up and down with every change

**Comment by gocarlos** [comment]
@jimsch please review this

**Comment by gocarlos** [comment]
I’d like to contribute some code, though it would make sense to format the code first

**Comment by jimsch** [comment]
As you said style is subjective.  I can live with most things but there are a couple of things that would drive me crazy:
*  Always use tabs to indent because different people think you should indent at different widths
*  In most cases I really want to have open braces NOT on a new line.  Function declarations are the big exception to that rule.

In short I did a lot of my learning of style with K&R.

**Comment by jimsch** [comment]
Part of the reason you are going to have a drop in coverage is that a large number of error cases don't have testing on them.  This means that chaning

~~~~
if (foo) deal with the error

to
if (foo) {
  deal with error
}
~~~
This increases the number of lines not covered.

**Comment by gocarlos** [comment]
> As you said style is subjective. I can live with most things but there are a couple of things that would drive me crazy:
> 
> * Always use tabs to indent because different people think you should indent at different widths
> * In most cases I really want to have open braces NOT on a new line.  Function declarations are the big exception to that rule.
> 
> In short I did a lot of my learning of style with K&R.

Sure, i’m 100% agnostic to the style.
Let me push a new version with this style

**Comment by gocarlos** [comment]
lets wait for https://github.com/cose-wg/COSE-C/pull/82 to land, otherwise we get tons of conflicts...

**Comment by gocarlos** [comment]
> * Always use tabs to indent because different people think you should indent at different widths

I have the feeling that spaces are the way to go with clang-format, when using tabs, the code looks pretty special (e.g. macros, function arguments etc.)...


It offers the option to costumize the width if using spaces, so not discussions needed here...
Also the default ones, LLVM, Chromium, Mozilla, Webkit etc. are all based on spaces, so I guess clang-format was tweaked for spaces... 

If you really want tabs, I'd suggest that you remove this last commit and play with the clang-format file as well as with the reference here: https://clang.llvm.org/docs/ClangFormatStyleOptions.html

**Comment by jimsch** [comment]
I can live with most of this, however some of it I really don't like.
1.  Please change to 4 character indents, this is why I use tabs not spaces because lots of people want different indent levels.
2.  I cannot find an option to control this, but I have found the following construct to be asking for errors
~~~
if (condition)
    statement;

vs
if (condition) {
    statement;
}
~~~~
It is just too easy to add a second statement in the if statement w/o adding the braces.   Is there a way to force doing braces in this case?
3.  Add BeforeCatch: false  and BeforeElse: false
4.  Do you know what the line length is?  I don't object to what is there I am just curious.

**Comment by gocarlos** [comment]
> Do you know what the line length is? I don't object to what is there I am just curious.

we are currently using chromium based style in general, this is based again on google style: 
https://google.github.io/styleguide/cppguide.html#Line_Length which has default line length of 80 chars

**Comment by gocarlos** [comment]
> 2\. I cannot find an option to control this, but I have found the following construct to be asking for errors

I'm not sure, if clang-format will be able to insert brakets to existing code, clang-format is mostly doing indentation, though we think exactly the same, and therefore we are using clang-tidy which reports on bad style: https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html

this could be added in a next MR

**Comment by gocarlos** [comment]
> > Do you know what the line length is? I don't object to what is there I am just curious.
> 
> we are currently using chromium based style in general, this is based again on google style:
> https://google.github.io/styleguide/cppguide.html#Line_Length which has default line length of 80 chars

this can be seen by running `./scripts/build_clang_tidy.sh` after applying  https://github.com/cose-wg/COSE-C/pull/84

**Comment by jimsch** [comment]
How soon do you think this is going to be ready to pull in.  I am waiting on some stuff that I want to do for it.

**Comment by gocarlos** [comment]
> How soon do you think this is going to be ready to pull in. I am waiting on some stuff that I want to do for it.

This can be done at any time, i can rebase today and let you review again?

**Comment by jimsch** [comment]
sounds good

**Comment by gocarlos** [comment]
> sounds good

just realized that it would be very good to merge https://github.com/cose-wg/COSE-C/pull/93 first, as otherwise we get conflicts with `cose_configure.h` and `cose_configure.h.in`

**Comment by gocarlos** [comment]
travis failing due to a timeout, nothing about this patch

**Pull request merged**